README
ExifTool by Phil Harvey (phil at owl.phy.queensu.ca)
----------------------------------------------------------------------------

ExifTool is a customizable set of Perl modules plus a full-featured
application for reading and writing meta information in a wide variety of
files, including the maker note information of many digital cameras by
various manufacturers such as Canon, Casio, FujiFilm, GE, HP, JVC/Victor,
Kodak, Leaf, Minolta/Konica-Minolta, Nikon, Olympus/Epson, Panasonic/Leica,
Pentax/Asahi, Reconyx, Ricoh, Samsung, Sanyo, Sigma/Foveon and Sony.

Below is a list of file types and meta information formats currently
supported by ExifTool (r = read, w = write, c = create):

  File Types
  ------------+-------------+-------------+-------------+------------
  3FR   r     | DYLIB r     | KEY   r     | OTF   r     | RWL   r/w
  3G2   r     | EIP   r     | LNK   r     | PAGES r     | RWZ   r
  3GP   r     | EPS   r/w   | M2TS  r     | PBM   r/w   | RM    r
  ACR   r     | ERF   r/w   | M4A/V r     | PDF   r/w   | SO    r
  AFM   r     | EXE   r     | MEF   r/w   | PEF   r/w   | SR2   r/w
  AI    r/w   | EXIF  r/w/c | MIE   r/w/c | PFA   r     | SRF   r
  AIFF  r     | EXR   r     | MIFF  r     | PFB   r     | SRW   r/w
  APE   r     | F4A/V r     | MKA   r     | PFM   r     | SVG   r
  ARW   r/w   | FFF   r/w   | MKS   r     | PGF   r     | SWF   r
  ASF   r     | FLA   r     | MKV   r     | PGM   r/w   | THM   r/w
  AVI   r     | FLAC  r     | MNG   r/w   | PICT  r     | TIFF  r/w
  BMP   r     | FLV   r     | MOS   r/w   | PMP   r     | TTC   r
  BTF   r     | FPX   r     | MOV   r     | PNG   r/w   | TTF   r
  CHM   r     | GIF   r/w   | MP3   r     | PPM   r/w   | VRD   r/w/c
  COS   r     | GZ    r     | MP4   r     | PPT   r     | VSD   r
  CR2   r/w   | HDP   r/w   | MPC   r     | PPTX  r     | WAV   r
  CRW   r/w   | HDR   r     | MPG   r     | PS    r/w   | WDP   r/w
  CS1   r/w   | HTML  r     | MPO   r/w   | PSB   r/w   | WEBP  r
  DCM   r     | ICC   r/w/c | MQV   r     | PSD   r/w   | WEBM  r
  DCP   r/w   | IDML  r     | MRW   r/w   | PSP   r     | WMA   r
  DCR   r     | IIQ   r/w   | MXF   r     | QTIF  r     | WMV   r
  DFONT r     | IND   r/w   | NEF   r/w   | RA    r     | X3F   r/w
  DIVX  r     | INX   r     | NRW   r/w   | RAF   r/w   | XCF   r
  DJVU  r     | ITC   r     | NUMBERS r   | RAM   r     | XLS   r
  DLL   r     | J2C   r     | ODP   r     | RAR   r     | XLSX  r
  DNG   r/w   | JNG   r/w   | ODS   r     | RAW   r/w   | XMP   r/w/c
  DOC   r     | JP2   r/w   | ODT   r     | RIFF  r     | ZIP   r
  DOCX  r     | JPEG  r/w   | OGG   r     | RSRC  r     |
  DV    r     | K25   r     | OGV   r     | RTF   r     |
  DVB   r     | KDC   r     | ORF   r/w   | RW2   r/w   |

  Meta Information
  ----------------------+----------------------+---------------------
  EXIF           r/w/c  |  CIFF           r/w  |  Ricoh RMETA    r
  GPS            r/w/c  |  AFCP           r/w  |  Picture Info   r
  IPTC           r/w/c  |  Kodak Meta     r/w  |  Adobe APP14    r
  XMP            r/w/c  |  FotoStation    r/w  |  MPF            r
  MakerNotes     r/w/c  |  PhotoMechanic  r/w  |  Stim           r
  Photoshop IRB  r/w/c  |  JPEG 2000      r    |  APE            r
  ICC Profile    r/w/c  |  DICOM          r    |  Vorbis         r
  MIE            r/w/c  |  Flash          r    |  SPIFF          r
  JFIF           r/w/c  |  FlashPix       r    |  DjVu           r
  Ducky APP12    r/w/c  |  QuickTime      r    |  M2TS           r
  PDF            r/w/c  |  Matroska       r    |  PE/COFF        r
  PNG            r/w/c  |  GeoTIFF        r    |  AVCHD          r
  Canon VRD      r/w/c  |  PrintIM        r    |  ZIP            r
  Nikon Capture  r/w/c  |  ID3            r    |  (and more)

See html/index.html for more details about ExifTool features.

ExifTool can be downloaded from

  Main server URL:  http://owl.phy.queensu.ca/~phil/exiftool/
  Alternate URL:    http://130.15.24.87/~phil/exiftool/

Note:  Please do not make links to the alternate URL above, because this
server is not permanent.

RUNNING

The exiftool script can be run right away without the need to install
Image::ExifTool.  For example, from within the exiftool directory you can
extract the information from one of the included test files by typing:

  ./exiftool t/images/ExifTool.jpg

If you move the exiftool script to a different directory, you must also
either move the contents of the lib directory or install the Image::ExifTool
package so the script can find the necessary libraries.

Note:  If you are using the Windows cmd shell, you may need to rename
'exiftool' to 'exiftool.pl' to run it directly from the command line.
Alternatively, you can run exiftool with the command 'perl exiftool'.

IF YOU ARE STILL CONFUSED

The exiftool script is a command line application.  You run it by typing
commands in a terminal window.  The first step is to determine the name of
the directory where you downloaded the ExifTool distribution package.
Assuming, for example, you downloaded it to a folder called "Desktop" in
your home directory, then you would type the following commands in a
terminal window to extract and run ExifTool:

  cd ~/Desktop
  gzip -dc Image-ExifTool-#.##.tar.gz | tar -xf -
  cd Image-ExifTool-#.##
  ./exiftool t/images/ExifTool.jpg

Note: You must replace "#.##" in the above commands with the actual version
number of ExifTool that you downloaded.  These commands extract meta
information from one of the test images.  To use one of your images instead,
enter the full path name of your file in place of "t/images/ExifTool.jpg".

INSTALLATION

You can install the Image::ExifTool package to make it available for use by
other Perl scripts by typing the following:

  perl Makefile.PL
  make
  make test
  make install

Notes:
  i) You need root access for the last step above.

  ii) Some Perl installations (like the standard OSX installation) may not
  contain the necessary files to complete the first step above.  But no
  worries:  You can install ExifTool manually by moving 'exiftool' and the
  'lib' directory to any directory in your current PATH (ie. /usr/bin).

  iii) In Windows, the "nmake" utility may be downloaded from Microsoft at
  http://support.microsoft.com/default.aspx?scid=kb;en-us;Q132084 -- Type
  "nmake" instead of "make" in the commands above if using this utility.

(Also see html/install.html for more help with installation.)

DEPENDENCIES

Requires Perl version 5.004 or later.  No other special libraries are
required, however the following modules are recommended for decoding
compressed and/or encrypted information from the indicated file types, and
for calculating digest values for some information types:

  Archive::Zip        (ZIP, DOCX, PPTX, XLSX, ODP, ODS, ODT, EIP and iWork)
  Compress::Zlib      (DNG, PNG, PDF, DCM, MIE and SWF files)
  Digest::MD5         (PDF files, IPTC information, and JPG Extended XMP)
  Digest::SHA         (PDF with AES-256 encryption)
  IO::Compress::Bzip2 (RWZ files)

COPYRIGHT AND LICENCE

Copyright 2003-2012, Phil Harvey

This is free software; you can redistribute it and/or modify it under the
same terms as Perl itself.

DISTRIBUTION FILES

Below is a list of the files/directories included in the full ExifTool
distribution package:

  Changes                   - Revision history
  ExifTool_config           - Sample run-time configuration file
  MANIFEST                  - Full list of distribution files
  META.yml                  - Standard CPAN dependency file
  Makefile.PL               - Makefile for installation
  README                    - This file
  exif2iptc.args            - Arguments for converting EXIF to IPTC
  exif2xmp.args             - Arguments for converting EXIF to XMP
  exiftool                  - The exiftool application (Perl script)
  gps2xmp.args              - Arguments for converting GPS to XMP
  gpx.fmt                   - Format file for creating GPX track
  gpx_wpt.fmt               - Format file for creating GPX waypoints
  html/                     - HTML documentation
  html/TagNames/            - HTML tag name documentation
  iptc2exif.args            - Arguments for converting IPTC to EXIF
  iptc2xmp.args             - Arguments for converting IPTC to XMP
  iptcCore.args             - Complete list of IPTC Core XMP tags
  kml.fmt                   - Format file for creating KML output files
  lib/                      - ExifTool Perl library modules
  pdf2xmp.args              - Arguments for converting PDF to XMP
  perl-Image-ExifTool.spec  - Red Hat Packaging Manager specification file
  t/                        - Verification test code
  t/images/                 - Verification test images
  xmp2exif.args             - Arguments for converting XMP to EXIF
  xmp2gps.args              - Arguments for converting XMP to GPS
  xmp2iptc.args             - Arguments for converting XMP to IPTC
  xmp2pdf.args              - Arguments for converting XMP to PDF

ADDITIONAL INFORMATION

Read the following files included in the full distribution for more
information:

  html/index.html           - Main ExifTool documentation
  html/install.html         - Installation instructions
  html/history.html         - Revision history
  html/ExifTool.html        - API documentation
  html/TagNames/index.html  - Tag name documentation
  html/geotag.html          - Geotag feature
  html/faq.html             - Frequently asked questions
  html/filename.html        - Renaming/moving files
  html/metafiles.html       - Working with metadata sidecar files
  html/struct.html          - Working with structured XMP information
  lib/Image/ExifTool/README - ExifTool library modules documentation

and if you have installed Image::ExifTool, you can also consult perldoc or
the man pages:

  perldoc exiftool
  perldoc Image::ExifTool
  perldoc Image::ExifTool::TagNames

  man exiftool
  man Image::ExifTool
  man Image::ExifTool::TagNames

Note: If the man pages don't work, it is probably because your man path is
not set to include the installed documentation.  See "man man" for
information about how to set the man path.

----------------------------------------------------------------------------


Changes
DO NOT EDIT THIS FILE -- it is generated from the html history files.

ExifTool Version History

RSS feed: http://owl.phy.queensu.ca/~phil/exiftool/rss.xml

Note: The most recent production release is Version 9.11. (Other versions are
considered development releases, and are not uploaded to CPAN.)

Jan. 2, 2013 - Version 9.11 (production release)

  - Improved decoding of some Sony tags
  - Changed 3 tag names to avoid a leading digit to fix XML validation problem
  - Fixed bug introduced in 9.04 that could double-encipher some Sony MakerNote
    information when writing (affected files are fixed by writing any tag with
    ExifTool 9.11)

Dec. 29, 2012 - Version 9.10

  - Added write support for a few new XMP-crs and XMP-photomech tags
  - Added a new Samsung LensType (thanks Jaroslav Stepanek)
  - Added a new Pentax LensType (thanks Helmut Schutz)
  - Added a new Canon LensType
  - Decode Sony A99 FocusMode (thanks Michael Tapes for the samples)
  - Tolerate (but warn about) up to 4 bytes of garbage at start of EXIF segment
  - Changed -P option to also preserve FileCreateDate on Windows (requires
    Win32API::File::Time)
  - Changed "[minor]" warning messages to capitalize the "M" (ie. "[Minor]") if
    processing is affected when the warning is ignored
  - Patched to avoid problem of slow processing with some corrupted EXIF

Dec. 15, 2012 - Version 9.09

  - Added a few new Google XMP GPano tags
  - Added a new Olympus CameraType
  - Added a couple of new Minolta LensTypes
  - Added two new Nikon LensID's (thanks David Puschel and Robert
    Rottmerhusen)
  - Decode Nikon D7000 AFPointsUsed and make this tag writable
  - Decode a new Olympus tag (thanks Christoph Anton Mitterer)
  - Renamed one of the FujiFilm RAF RawImageWidth/Height pairs to
    RawImageFullWidth/Height
  - Changed -stay_open when combined with -q to flush output after each command
    (as already done without -q) (requires IO::Handle)
  - Fixed problem shifting FileCreateDate when writing other "real" tags in the
    same command

Nov. 26, 2012 - Version 9.08

  - Fixed bug introduced in 9.07 that broke writing of FileModifyDate

Nov. 24, 2012 - Version 9.07

  - Added ability to read/write FileCreateDate (Windows only)
  - Added ability to read FileInodeChangeDate (non-Windows only)
  - Added support for new tags in DNG 1.4 specification
  - Added support for Google Photosphere GPano XMP tags
  - Added a couple of new Olympus filter effects
  - Changed a Panasonic LensType (thanks Olaf Ulrich)
  - API Changes:
      - Enhanced SetFileModifyDate() to write FileCreateDate (Windows only)

Nov. 17, 2012 - Version 9.06

  - Added support for Nikon maker notes in images from any camera make (as
    written by Capture NX2)
  - Added support for FujiFilm X-E1 RAF images
  - Added a new Olympus CameraType
  - Added a new PentaxModelID and a new Pentax LensType
  - Extract FileCreateDate (Windows) and FileInodeChangeDate (other systems)
  - Fixed bug decoding UTF-16 ID3 synchronized lyrics

Nov. 10, 2012 - Version 9.05

  - Added ability to read APE metadata from MP3 audio files
  - Decode ID3 synchronized lyrics/text information
  - Decode maker notes in Leica V-LUX40 MP4 videos
  - Decode Sony A99 AFPointSelected (thanks Michael Tapes for the samples)
  - Improved decoding of some Sony tags (thanks Jos Roost)
  - API Changes:
      - Removed GeoNoInterpolate option (just set GeoMaxIntSecs to 0 instead)

Nov. 3, 2012 - Version 9.04 (production release)

  - Added two new Sony LensType values (thanks Matthias Paul)
  - Added a few new Canon LensType values
  - Added a couple of new PentaxModelID's and decode some new K-5 II values
  - Added support for some new XMP tags written by the Apple iPhone 5
  - Added a new Olympus CameraType
  - Decode more Sony tags/values (thanks Jos Roost)
  - Decode Nikon HDRInfo (thanks Stefan)
  - Decode some FlashInfo tags for new Nikon models
  - Decode a few WM ID3 tags (some documentation on these would be nice)
  - Fixed bug which could cause truncated/garbage ID3v2 strings to be returned
  - Fixed -globalTimeShift option to also work when copying tags
  - Fixed decoding of Nikon AFFineTuneAdj for FirmwareVersion 1.10B (thanks
    Michael Tapes for the samples for this and the A77)
  - Fixed problem where a few tags (FileSequence, NewGUID and Now) were not
    available for use with the -p option
  - API Changes:
      - Added RequestAll and GeoNoInterpolate options
      - Fixed problem in SetNewValue when setting the Raw value of some tags

Oct. 13, 2012 - Version 9.03

  - Added new feature to provide control over directory levels in %d strings
  - Added ability to write OtherImage in NEF images
  - Added a new Pentax LensType
  - Added a few new CanonModelID's (thanks Laurent Clevy)
  - Added a new Nikon LensID (thanks Geert De Soete)
  - Added a few new Olympus CameraType values
  - Decode some new CameraInfo tags for the Canon EOS 650D
  - Decode a number of new Sony tags (thanks Jos Roost)
  - Improved decoding of some Sigma tags for the DP1/DP2 Merrill
  - Give priority to EXIF tags over SigmaRaw tags X3F images
  - Changed Samsung lens names to include "NX" (thanks Jaroslav Stepanek)
  - Fixed misleading verbose "TAG is not writable" messages when copying
    list-type tags
  - API Changes:
      - Enhanced GetValue() to allow return of 'Rational' value

Sept. 6, 2012 - Version 9.02

  - Added a new Nikon LensID (thanks Joseph Heled)
  - Added a new EXIF SubFileType value used in DNG images
  - Added write support for Apple Adjustment Settings XMP tags (XMP-aas)
  - Added a couple of new Samsung LensType values (thanks Jaroslav Stepanek)
  - Added a couple of new Canon LensType values and a new CanonModelID
  - Decode a number of new Sony tags (thanks Jos Roost)
  - Enhanced "-o -" feature to allow output file type to be specified
  - Extract last file access time as FileAccessDate
  - Allow tags to be set from files which are zero bytes in size

lib/Image/ExifTool/BMP.pm
#------------------------------------------------------------------------------
# File:         BMP.pm
#
# Description:  Read BMP meta information
#
# Revisions:    07/16/2005 - P. Harvey Created
#
# References:   1) http://www.fortunecity.com/skyscraper/windows/364/bmpffrmt.html
#               2) http://www.fourcc.org/rgb.php
#------------------------------------------------------------------------------

package Image::ExifTool::BMP;

use strict;
use vars qw($VERSION);
use Image::ExifTool qw(:DataAccess :Utils);

$VERSION = '1.04';

# BMP chunks
%Image::ExifTool::BMP::Main = (
    PROCESS_PROC => \&Image::ExifTool::ProcessBinaryData,
    GROUPS => { 2 => 'Image' },
    NOTES => q{
        There really isn't much meta information in a BMP file as such, just a bit
        of image related information.
    },
    # 0 => size of bitmap structure:
    #        12  bytes => 'OS/2 V1',
    #        40  bytes => 'Windows V3',
    #        64  bytes => 'OS/2 V2',
    #        68  bytes => some bitmap structure in AVI videos
    #        108 bytes => 'Windows V4',
    #        124 bytes => 'Windows V5',
    4 => {
        Name => 'ImageWidth',
        Format => 'int32u',
    },
    8 => {
        Name => 'ImageHeight',
        Format => 'int32u',
    },
    12 => {
        Name => 'Planes',
        Format => 'int16u',
    },
    14 => {
        Name => 'BitDepth',
        Format => 'int16u',
    },
    16 => {
        Name => 'Compression',
        Format => 'int32u',
        # (formatted as string[4] for some values in AVI images)
        ValueConv => '$val > 256 ? unpack("A4",pack("V",$val)) : $val',
        PrintConv => {
            0 => 'None',
            1 => '8-Bit RLE',
            2 => '4-Bit RLE',
            3 => 'Bitfields',
            4 => 'JPEG', #2
            5 => 'PNG', #2
            # pass through ASCII video compression codec ID's
            OTHER => sub {
                my $val = shift;
                # convert non-ascii characters
                $val =~ s/([\0-\x1f\x7f-\xff])/sprintf('\\x%.2x',ord $1)/eg;
                return $val;
            },
        },
    },
    20 => {
        Name => 'ImageLength',
        Format => 'int32u',
    },
    24 => {
        Name => 'PixelsPerMeterX',
        Format => 'int32u',
    },
    28 => {
        Name => 'PixelsPerMeterY',
        Format => 'int32u',
    },
    32 => {
        Name => 'NumColors',
        Format => 'int32u',
        PrintConv => '$val ? $val : "Use BitDepth"',
    },
    36 => {
        Name => 'NumImportantColors',
        Format => 'int32u',
        PrintConv => '$val ? $val : "All"',
    },
);

#------------------------------------------------------------------------------
# Extract EXIF information from a BMP image
# Inputs: 0) ExifTool object reference, 1) dirInfo reference
# Returns: 1 on success, 0 if this wasn't a valid BMP file
sub ProcessBMP($$)
{
    my ($exifTool, $dirInfo) = @_;
    my $raf = $$dirInfo{RAF};
    my $buff;

    # verify this is a valid BMP file
    return 0 unless $raf->Read($buff, 14) == 14;
    return 0 unless $buff =~ /^BM/;
    return 0 unless $raf->Read($buff, 40) == 40;
    SetByteOrder('II');
    return 0 unless Get32u(\$buff, 0) >= 40;    # make sure info is at least 40 bytes
    $exifTool->SetFileType();   # set the FileType tag
    my %dirInfo = (
        DataPt => \$buff,
        DirStart => 0,
        DirLen => length($buff),
    );
    my $tagTablePtr = GetTagTable('Image::ExifTool::BMP::Main');
    $exifTool->ProcessDirectory(\%dirInfo, $tagTablePtr);
    return 1;
}

1;  # end

__END__

=head1 NAME

Image::ExifTool::BMP - Read BMP meta information

=head1 SYNOPSIS

This module is used by Image::ExifTool

=head1 DESCRIPTION

This module contains definitions required by Image::ExifTool to read BMP
(Windows Bitmap) images.

=head1 AUTHOR

Copyright 2003-2008, Phil Harvey (phil at owl.phy.queensu.ca)

This library is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=head1 REFERENCES

=over 4

=item L<http://www.fortunecity.com/skyscraper/windows/364/bmpffrmt.html>

=back

=head1 SEE ALSO

L<Image::ExifTool::TagNames/BMP Tags>,
L<Image::ExifTool(3pm)|Image::ExifTool>

=cut

